lib/instrumentation/span-compression.js

```diff
@@ -0,0 +1,176 @@
+'use strict'
+const STRATEGY_EXACT_MATCH = 'exact_match'
+const STRATEGY_SAME_KIND = 'same_kind'
+
+class SpanCompression {
+  constructor (agent) {
+    this._bufferedSpan = null
+    this._agent = agent
+    this.timestamp = null
+    this.duration = null
+    this.composite = {
+      count: 0,
+      sum: 0,
+      compression_strategy: null
+    }
+  }
+
+  setBufferedSpan (span) {
+    this._bufferedSpan = span
+  }
+
+  getBufferedSpan () {
+    return this._bufferedSpan
+  }
+
+  // Compares two spans and returns which compression strategy to use
+  // or returns false if the second span can't be compressed into the
+  // first.
+  //
+  // @param Span compositeSpan
+  // @param Span toCompressSpan
+  // @returns boolean|String
+  _getCompressionStrategy (compositeSpan, toCompressSpan) {
+    if (!this._isEnabled() || !compositeSpan._destination || !toCompressSpan._destination) {
+      return false
+    }
+
+    const isSameKind = compositeSpan.type === toCompressSpan.type &&
+    compositeSpan.subtype === toCompressSpan.subtype &&
+    compositeSpan._destination.service.resource === toCompressSpan._destination.service.resource
+
+    if (!isSameKind) {
+      return false
+    }
+
+    let strategy = STRATEGY_SAME_KIND
+    if (compositeSpan.name === toCompressSpan.name) {
+      strategy = STRATEGY_EXACT_MATCH
+    }
+
+    // check duration limits
+    if (!this._checkDuration(compositeSpan._duration, toCompressSpan._duration, strategy)) {
+      return false
+    }
+
+    return strategy
+  }
+
+  // Sets initial composite values or confirms strategy matches
+  //
+  // Returns true if spanToCompress can be compressed into compositeSpan,
+  // returns false otherwise.
+  //
+  // @param Span compositeSpan
+  // @param Span spanToCompress
+  // @returns boolean
+  _initCompressionStrategy (compositeSpan, spanToCompress) {
+    if (!this.composite.compression_strategy) {
+      // If no strategy is set, check if strategizable or not. If so,
+      // set initial values.  If not, bail.
+      this.composite.compression_strategy = this._getCompressionStrategy(
+        compositeSpan,
+        spanToCompress
+      )
+      if (!this.composite.compression_strategy) {
+        return false
+      }
+
+      // set initial composite context values
+      this.timestamp = compositeSpan.timestamp
+      this.composite.count = 1
+      this.composite.sum = compositeSpan._duration
+    } else {
+      // if so, compare with the compression strat and bail if mismatch
+      const strat = this._getCompressionStrategy(compositeSpan, spanToCompress)
+      if (strat !== this.composite.compression_strategy) {
+        return false
+      }
```

Comment:
I disagree (or think I do) -- this appears to work correctly. Both the _test tryToCompress exact match_ and _test tryToCompress same kind_ include tests for "the strategy changed".  I also added a new test named _test tryToCompress same kind, then exact match_  with an additional scenario to confirm that things work as intended. 

Once a span becomes a same kind composite its name is going to be prefixed with the string _Calls to_ which means the names will no longer exact match. 

If I've misunderstood could you express the scenario you're talking about as a test case?

Reply:
Okay, understood. Here is a contrived example (creating a span with the name "Calls to mysql") that can break this, but seems unlikely :)

```js
var apm = require('./').start({
  metricsInterval: '0s',
  centralConfig: false,
  cloudProvider: 'none',
  spanCompressionEnabled: true,
  spanCompressionSameKindMaxDuration: '50ms'
})

function main () {
  var t0 = apm.startTransaction('t0')
  setImmediate(() => {
    var s1 = t0.startSpan('s1', 'db', 'mysql')
    s1.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
    setTimeout(() => {
      s1.end()
      var s2 = t0.startSpan('s2', 'db', 'mysql')
      s2.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
      setTimeout(() => {
        s2.end()

        // Create a span with a contrived `name` to fool `SpanCompression._initCompressionStrategy()`
        var s3 = t0.startSpan('Calls to mysql', 'db', 'mysql')
        s3.setDestinationContext({ service: { name: 'mysql', resource: 'mysql', type: 'db' } })
        setTimeout(() => {
          s3.end()
          t0.end()
        }, 10)
      }, 10)
    }, 10)
  })
}

main()
```

This results in:

```
    trace 2c2404
    `- transaction 004fa5 "t0" (41.486ms, outcome=unknown)
       `- span fb8e47 "Calls to mysql" (24.776ms, mysql, compressed 2 same_kind spans sum 23.366ms)
       `- span 7e5a8e "Calls to mysql" (11.518ms, mysql)
```

Reply:
Unresolving this in case you didn't see my follow-up comment. Feel free to resolve again if you don't think the contrived case is worth guarding against.